Feat/timezone IANA support

nikosalonen opened the pull request:
This PR introduces better, DST aware timezone support mentioned in the issue #78 
The timezones are still hardcoded and it would be better to have a proper API for those.


- **New Features**
  - IANA timezone support: account timezone selection, signup auto-detection, dashboard migration banner with suggested zones, tracking script/embeds accept data-timezone, and updated dashboard/client helpers.

- **Bug Fixes**
  - Timezone-aware timestamps across tracking, trackpage, queries, reporting and exports; dumps include timezone/offset; dashboard uses server-provided precise offsets.

- **Documentation**
  - Integration guides and platform docs updated with timezone examples, migration steps, and platform-specific instructions.

- **Tests**
  - Added some tests
  
  
  Please test this to see that everything works in your end. I've also fixed many linter errors.

nikosalonen committed:
feat(models): add IANA timezone helpers on User (validate and store)

nikosalonen committed:
refactor(user): improve code formatting and simplify for loop syntax

nikosalonen committed:
feat(api): accept and persist timezone in account edit and register

nikosalonen committed:
feat(time): compute user-local time via IANA timezone; refactor GetVisits and dump

nikosalonen committed:
feat(time): prefer stored timezone, then stored utcoffset, then request utcoffset

nikosalonen committed:
feat(dump): include meta.offsetMinutesNow and meta.timezone in dump payload

nikosalonen committed:
fix(dump): reorder import statements for consistency

nikosalonen committed:
feat(tracking): add optional data-timezone to embed and send it from scripts

nikosalonen committed:
fix(dashboard): correct broken syntax and variable declarations

nikosalonen committed:
feat(dashboard): prefer server-provided offsetMinutesNow over client getUTCOffset()

nikosalonen committed:
refactor(load): improve script loading syntax and formatting for clarity

nikosalonen committed:
refactor(counter): improve code formatting and simplify logic for date calculations

nikosalonen committed:
refactor(graph): improve variable naming and use arrow functions for better readability

nikosalonen committed:
refactor(time): improve code formatting and use block syntax for clarity in sumHours method

nikosalonen committed:
refactor(dashboard): enhance code clarity by using block syntax and simplifying date range calculation

nikosalonen committed:
refactor(dashboard): streamline data handling and enhance date range calculation logic

nikosalonen committed:
refactor(dashboard): optimize date range calculation using Array.from for improved readability

nikosalonen committed:
fix(graph): resolve lint errors - remove unused params, fix variable declarations, add missing return

nikosalonen committed:
feat(dashboard): complete server-provided offset integration in getDumpURL

nikosalonen committed:
feat(ui): replace UTC offset dropdown with IANA timezone picker in account settings

nikosalonen committed:
feat(timezones): expand to comprehensive IANA timezone list with 70+ options covering all UTC offsets and DST variations

nikosalonen committed:
fix(editaccount): remove unnecessary whitespace in timezone definitions for cleaner code

nikosalonen committed:
feat(migration): implement user timezone migration from utcoffset to IANA timezones

- Add SuggestTimezoneFromOffset mapping function
- Include migration suggestions in user dump when needed
- Create timezone-migration component with suggestion UI
- Add automatic timezone update and manual selection options
- Include dismissal functionality and account settings integration
- Add responsive styling for migration banner

nikosalonen committed:
fix: remove unnecessary whitespace in dump.go and user.go for cleaner code

nikosalonen committed:
feat(archive): implement timezone-aware archive range calculations

- Fix QueryArchive date filtering bug (missing query assignments)
- Update query endpoint to parse dates in user timezone instead of UTC
- Enable frontend date functions for timezone-aware client calculations
- Use server-provided offset in getUTCNow for consistency
- Ensure archive date ranges use timezone-aware boundaries

nikosalonen committed:
refactor(dashboard): improve code readability and consistency in selector component

nikosalonen committed:
refactor(dashboard): enhance code readability by using const and consistent formatting in pie.js

nikosalonen committed:
fix(archive): update QueryArchiveOldestDate to handle nil date values and improve error handling

nikosalonen committed:
refactor(dashboard): expose utility functions as global by assigning to window object

nikosalonen committed:
test: add comprehensive timezone tests for DST transitions and fractional offsets

- Test DST transitions for US/Eastern, Europe/Berlin, and Australia/Sydney
- Test fractional timezone offsets (half-hour and quarter-hour)
- Test timezone-aware date parsing and UTC conversions
- Test IANA timezone resolution and validation
- Test timezone migration suggestions based on UTC offset
- Test DST boundary conditions around spring/fall transitions
- Cover edge cases like Chatham Island (+13:45 DST), Nepal (+5:45), Iran (+3:30)
- Validate server-side offset calculations match expected values

nikosalonen committed:
docs: update documentation to prefer timezone over utcoffset

- Update external tracking script blog post with timezone examples
- Modify tracking code component to prefer timezone parameter
- Add comprehensive timezone configuration guide to integration.md
- Include migration guide from utcoffset to timezone with common conversions
- Update all platform-specific guides (WordPress, Shopify, Squarespace, Wix, Weebly)
- Add timezone feature highlight to README.md
- Maintain backward compatibility with utcoffset for existing users
- Emphasize benefits: automatic DST handling, fractional offsets, accuracy

nikosalonen committed:
refactor(navbar): improve code formatting and readability in navbar.js

nikosalonen committed:
fix(navbar): correct variable declaration for text in connectedCallback

nikosalonen committed:
refactor(pwyw): improve code formatting and readability in pwyw.js

nikosalonen committed:
fix: remove unused utcoffset parameter and fix lint issues

- Remove unused utcoffset parameter from dashboard-dynamics component
- Update dashboard.js connectData call to match component signature
- Convert string concatenation to template literals
- Change let to const for immutable variables
- Fix arrow function syntax in forEach callbacks

nikosalonen committed:
test: add comprehensive timezone tests for various scenarios

- Introduce tests for DST transitions across multiple timezones including US/Eastern, Europe/Berlin, and Australia/Sydney.
- Validate fractional timezone offsets and ensure correct parsing of timezone-aware dates.
- Implement tests for timezone resolution and migration suggestions based on UTC offsets.
- Cover edge cases and boundary conditions related to DST transitions.

nikosalonen committed:
refactor(app): enhance file serving logic and security in NewApp function

- Introduce a file server handler to securely serve files based on host configuration.
- Remove redundant MIME type setting code, leveraging http.FileServer for automatic Content-Type handling.
- Improve code readability and maintainability by restructuring file serving logic.

nikosalonen committed:
fix(user): enhance timezone handling in CurrentLocation method

- Update CurrentLocation to fall back to UTC when the stored timezone is invalid.
- Improve error handling by ensuring a valid timezone is returned or defaulted to UTC.

nikosalonen committed:
fix(test): correct DST transition date for Europe/Berlin test case

nikosalonen committed:
fix(script): conditionally add utcoffset to params if defined

nikosalonen committed:
fix(navbar): store user dump in savedUserDump on push-navbar-dump event

nikosalonen committed:
fix(dashboard): add input validation to percentRepr function to prevent invalid calculations

nikosalonen committed:
fix(dashboard): sort available dates correctly using Date objects

nikosalonen committed:
fix(setup): enhance draw function to include timezone handling

nikosalonen committed:
fix(utils): correct typo in modal notify display style

nikosalonen committed:
fix(welcome): correct typos in modal recovery notification message

nikosalonen committed:
fix(timezone): implement timezone preference handling and improve registration process

- Prioritize IANA timezone for new users while maintaining backwards compatibility with UTC offset.
- Update welcome page to reflect timezone detection and provide user-friendly information.
- Introduce utility functions for timezone detection and information retrieval.

nikosalonen committed:
fix(dashboard): improve timezone offset handling by using precise minutes

- Update dashboard data connections to prefer precise minute offsets over legacy hour calculations.
- Ensure backward compatibility by maintaining legacy hour parameters for UI.
- Refactor offset calculations to enhance clarity and maintainability.

nikosalonen committed:
fix(dashboard): handle undefined archives in addArchivesToDump function

nikosalonen committed:
fix(utils): correct typo in error callback parameter name

nikosalonen committed:
feat(app): implement noDirectoryListingFS to prevent directory listings in file server

nikosalonen committed:
fix(dashboard): adjust timezone preference handling order in data connection

nikosalonen committed:
fix(navbar): ensure eventSourceObj is closed only if open to save resources

nikosalonen committed:
fix(register): return early on invalid timezone to prevent further processing

nikosalonen committed:
fix(app): enhance noDirectoryListingFS to serve index.html for directories

nikosalonen committed:
fix(dashboard): add guard clause for empty or undefined myDates in dFillDatesToNow function

nikosalonen committed:
fix(utils): normalize timezone detection to handle undefined or non-string values

nikosalonen committed:
fix(dashboard): update getUTCNow function to use UTC for accurate time calculations

nikosalonen committed:
fix(dashboard): compute day ranges in UTC to prevent DST-related issues in dFillDatesToNow function

nikosalonen committed:
fix(dashboard): correct tooltip value retrieval in chart configuration

nikosalonen committed:
refactor(dashboard): change variable declarations to const for selector and allConnectedData

nikosalonen committed:
fix(dashboard): update offset calculation to use getClientOffsetMinutes for improved accuracy

nikosalonen committed:
fix(dashboard): enhance numberFormat function to handle non-numeric inputs and improve internationalization support

nikosalonen committed:
fix(dashboard): normalize date grouping to UTC to prevent local timezone drift

nikosalonen committed:
refactor(app): remove noDirectoryListingFS and handle directory requests directly in NewApp

nikosalonen committed:
refactor(timezone_test): simplify DST condition checks using switch statement

nikosalonen committed:
fix(user): correct initialization of PubSubConn to use named field for clarity

nikosalonen committed:
fix(user): update error message for user retrieval to follow consistent formatting

nikosalonen committed:
refactor(user): simplify CurrentLocation function by removing redundant timezone loading logic

nikosalonen committed:
refactor(timezone): update timezone mappings and test cases to include new fractional timezones and improve accuracy

nikosalonen committed:
fix(account): escape utcoffset value in hidden input for improved security

nikosalonen committed:
feat(account): add automatic timezone detection fallback for user accounts

nikosalonen committed:
fix(account): clear utcoffset input when timezone selection changes

nikosalonen committed:
refactor(editaccount): add ARIA attributes for improved accessibility in timezone selection

nikosalonen committed:
fix(editaccount): correct typo in function name for show/hide preferred sites

nikosalonen committed:
fix(editaccount): change button to type="button" for better semantics

nikosalonen committed:
refactor(editaccount): update timezone labels for clarity and consistency

ihucos commented:
Hello nikosalon,

thank you very much for the PR.

It's adding about 3500 lines and removing almost 2000 lines. In order to review it we would need to get into the core of it and try to understand the approach and actual feature. I have some questions:

* Can you explain the main approach and the handling of the integrations that become legacy after this PR. One thought is that maybe it is actually possible to migrate previous integrations somehow by guessing the timezone from the utc offset, domain ending, webpage language, visitors geolocation etc. Then maybe we could avoid having something declared as legacy in the code. Just a though and maybe out of scope for this PR, could also be done as a separate step.

* Can you explain me which AI tools you used? I'd be curious. The many commits look like Devin. I'd need to know which inputs to this PR comes from AI and which ones from you as a human.

* We have a typical problem with AI tools which is "overshooting". I would for example not update previous blog posts. I would also ask you to only change things specific to the feature at hand in order to make the reviewing seamless. 

* Which formatting tools are you using? I like the formating. In the moment the project uses `make format`. Hmmm maybe we can see better formating as a separate problem?

Thanks and cheers

ihucos

nikosalonen committed:
fix(accountedit): return early on invalid timezone to prevent further processing

nikosalonen committed:
fix(dump): ensure needsTimezoneUpdate is set correctly based on user timezone suggestions

nikosalonen committed:
refactor(register): encapsulate UTC offset handling in a helper function for clarity and maintainability

nikosalonen committed:
refactor(user): enhance timezone offset mapping to include fractional offsets and clarify input handling

nikosalonen committed:
run format

nikosalonen committed:
fix(dashboard): ensure proper handling of timezone offset by converting to Number for consistency

nikosalonen committed:
fix(dashboard): update getUTCNow to return formatted date string for consistency with daysRange expectations

nikosalonen committed:
refactor(load): replace forEach with for...of for improved iteration over script files

nikosalonen committed:
refactor(welcome): implement safe DOM manipulation for timezone registration and improve user feedback

nikosalonen committed:
revert: blog post beyond-the-hype back to master state

nikosalonen committed:
revert: blog and page templates back to master state

nikosalonen committed:
revert: README features addition out of PR scope

nikosalonen committed:
fix(dashboard): sanitize visits parsing and asset slugs; harden referrer link

nikosalonen committed:
feat(user): add EnsureSiteLink method to verify site existence without incrementing visit counter

nikosalonen committed:
fix(dashboard): enhance referrer link validation to ensure proper URL protocol handling

nikosalonen committed:
fix(user): update EnsureSiteLink method to return error on failure

nikosalonen committed:
feat(makefile): add timezone generation target

nikosalonen committed:
feat(editaccount): implement dynamic timezone loading with offsets and improve timezone display

nikosalonen committed:
feat(timezones): add script to generate a sorted list of timezones with offsets and aliases

nikosalonen committed:
refactor(editaccount): remove hardcoded timezone list, initializing TIMEZONES as an empty array

nikosalonen committed:
refactor(editaccount): standardize string quotes and improve readability in timezone loading logic

nikosalonen committed:
feat(timezones): generate timezone mappings from JSON and update Makefile for automation

nikosalonen committed:
feat(timezones): implement lazy loading of timezone data and enhance timezone formatting with current offsets

nikosalonen committed:
feat(timezones): enhance timezone data structure with displayId and continent fields for improved formatting

nikosalonen committed:
refactor(timezone-migration): simplify fallback logic for timezone display by removing hardcoded city mapping

nikosalonen committed:
feat(makefile): update timezone generation to use configurable TZDB_VERSION and improve error handling

nikosalonen committed:
fix(timezone_test): update test cases for UTC-5 and UTC+1 to reflect accurate timezone suggestions

nikosalonen committed:
refactor(selector): optimize jQuery selectors for range selection updates

nikosalonen committed:
fix(settings): add null check for counter-trackingcode before upgrading and drawing

nikosalonen committed:
refactor(timezone-migration): update timezone migration modal structure and improve suggestion rendering

nikosalonen committed:
refactor(timezone-migration, selector): standardize string quotes and improve jQuery selector readability

nikosalonen committed:
feat(dashboard): add timezone handling to user preferences for improved tracking

nikosalonen committed:
fix(makefile): improve error handling in timezone generation and streamline formatting commands

nikosalonen committed:
refactor(gen_timezones_go): enhance timezone sorting and grouping logic for improved data accuracy

nikosalonen committed:
fix(timezone-migration): add dismissal check for timezone migration modal and improve timezone offset handling

nikosalonen committed:
feat(timezone-migration): enhance timezone suggestion display with "Show more" functionality and improve continent-based selection

nikosalonen committed:
fix(timezone-migration): ensure proper HTML escaping in timezone suggestion buttons

nikosalonen commented:
Hi @ihucos 

Sorry, I completely missed the `make format` command. I very rarely work with Makefiles and didn't think to look there. I now ran it and the changes aren't that big anymore. Some files still seem to have indentation changes, which could be because of different versions of prettier maybe? I usually use [Biome](https://biomejs.dev/) for format and lint, and many small changes you can find in the code are because of Biome's linter warnings.

I also reverted all the "overshooting" I could find. Sorry about that.

I'm using [Cursor](https://cursor.com/) editor and [Coderabbit AI](https://coderabbit.ai/) for code quality. I cannot tell what code is by me and what is by AI as my workflow is very integrated with the tools. I did create the update plan on my own though.

This PR adds IANA timezone support without breaking any existing UTC-offset integrations. The tracking script now accepts `data-timezone="Area/City"`, the user preference stores an IANA name, and time calculations prefer that value whenever it is present. If a timezone has not been set, the system continues to honor the legacy `data-utcoffset` and falls back to UTC plus the configured offset.

Time resolution is handled in `backend/lib/ctx.go` by `UserNow(user models.User)`, which prefers the `timezone` preference and otherwise falls back to `utcoffset`. The user's timezone preference is managed in `backend/models/user.go` via `SetTimezone`, `ReadTimezone`, and `CurrentLocation`. Suggested IANA zones for a given offset come from the `offsetToTimezones` map and `SuggestTimezoneFromOffset` in the same file. Whether a user should be prompted to upgrade is determined by `NeedsTimezoneUpdate` (true when `utcoffset` exists but `timezone` is empty). On the frontend, `static/components/base/timezone-migration.js` renders the upgrade prompt. The tracking snippet in `static/components/counter-trackingcode.js` accepts both `data-timezone` and the legacy `data-utcoffset`. Behavior is covered by tests in `backend/timezone_test.go`, including DST and fractional offsets.

The timezone picker is now driven by a generated JSON file instead of a hardcoded list. A Makefile target (`gen-timezones`) fetches the Noda Time TZDB JSON for a pinned version and produces `static/timezones.json` using `scripts/timezones.jq`. The output includes canonical IANA IDs and alias entries, and is pre-sorted by UTC offset and then by name. The account settings component `static/components/base/editaccount.js` loads this JSON, builds labels that display the current offset (for example, `UTC+02 (CEST) — Europe/Berlin`), and falls back gracefully if the file is unavailable. The previous hardcoded `TIMEZONES` array has been removed.

Nothing is marked as legacy in the codebase. Documentation recommends using IANA timezones (`data-timezone`) while explicitly stating that UTC-offset integrations continue to be supported. The UTC-offset approach does not handle daylight saving time and has limitations with fractional offsets, which is why the IANA path is recommended going forward.

When `NeedsTimezoneUpdate` is true, the UI shows a migration banner. The banner, implemented in `timezone-migration.js`, presents suggested IANA zones derived from `SuggestTimezoneFromOffset`. Users can accept a suggestion (which persists the choice via `SetTimezone`), pick a different zone manually, or dismiss the prompt and continue with UTC-offset behavior. No migration is forced. The scripts use `data-utcoffset` until the timezone is set. After that, the scripts will have the `data-timezone` value.


All existing embeds continue working as they are, and no changes are required. 


<img width="526" height="513" alt="image" src="https://github.com/user-attachments/assets/91bd4b56-eb2d-44db-be31-25cb442e5a17" />


<img width="907" height="726" alt="image" src="https://github.com/user-attachments/assets/598b4c5f-0f5d-4e99-b81c-db96ff9396b1" />

<img width="498" height="245" alt="image" src="https://github.com/user-attachments/assets/d8462d8d-fd28-4a07-a118-db041f443f35" />

<img width="582" height="459" alt="image" src="https://github.com/user-attachments/assets/5020bc98-90e2-416e-84eb-732758ed96f1" />



### Reviewer links to code

- Time resolution fallback (IANA preferred, UTC offset otherwise): [`backend/lib/ctx.go` `UserNow`](https://github.com/ihucos/counter.dev/pull/135/files#diff-4551bdb9d3230d834a9b89b482c11a1fd228c6076cdf077019ce8f70398a5045)
- User timezone preference APIs: [`backend/models/user.go` `SetTimezone`/`ReadTimezone`/`CurrentLocation`](https://github.com/ihucos/counter.dev/pull/135/files#diff-e8308f4622b4f2ea396681c64db6129e2809afbd0ebe5f0f7c2948a29176ab0b)
- Offset → IANA suggestions map and helper: [`backend/models/user.go` `offsetToTimezones` and `SuggestTimezoneFromOffset`](https://github.com/ihucos/counter.dev/pull/135/files#diff-e8308f4622b4f2ea396681c64db6129e2809afbd0ebe5f0f7c2948a29176ab0b)
- Migration trigger flag: [`backend/models/user.go` `NeedsTimezoneUpdate`](https://github.com/ihucos/counter.dev/pull/135/files#diff-e8308f4622b4f2ea396681c64db6129e2809afbd0ebe5f0f7c2948a29176ab0b)
- Add migration suggestions and expose current offset minutes to frontend: [`backend/endpoints/dump.go` (suggestions and offsetMinutesNow meta)](https://github.com/ihucos/counter.dev/pull/135/files#diff-bf8c5cfee583fffa01ac741af451c010af8945129c1bb2566546b5221220c9d0)
- Dashboard uses precise minute offsets and sends them to backend: [`static/js/dashboard.js`](https://github.com/ihucos/counter.dev/pull/135/files#diff-660e4b5f235bfe167a72cb786796570ed22bae39bf9b1200c046064489bc581b)
- Timezone selector loads generated list and renders offsets: [`static/components/base/editaccount.js`](https://github.com/ihucos/counter.dev/pull/135/files#diff-8716b5a2076ea77fa37358aa8240135130223edfec7c84a84656eab124a196dd)

ihucos commented:
Hello nikosalonen,

thanks for explaining it it to me and reducing the number of lines changed.

I have thought about the PR and think that it is too big. I feel I cant get in 6000+ lines of code in without guaranteeing a good reliability for the existing users. What I would suggest is to try to achieve something similar with less changes. I do believe it can be done.

## Some ideas
* We can do a data migration to convert the utcoffset to a matching timezone. Then optionally notify users. I think we don't need the fancy "select new timezone now or skip" screen. Of course that is UX vs complexity compromise. But then we can drop the utcoffset "column" in redis at once, which further reduces complexity.

* A way to use timezones right now would be to let the client calculate the utc offset based on an actual timezone. 
```
# this lives in the tracking script
let utc_offset_parameter = (
  new Date(
    new Date().toLocaleString("en-US", { timeZone: "Europe/Berlin" })
  )
  -
  new Date(
    new Date().toLocaleString("en-US", { timeZone: "UTC" })
  )
) / 36e5
```
Then we would also need to test if floating point utc offsets actually work. Not sure, just an idea how to avoid complexity.


## There is an issue with the current approach
Unfortanly we can't have a roundtrip to the redis server when registering unique pageviews or pages. We can send things to redis but if we wait for the response, then send sth again to redis, the performance degrades. In the moment I am getting something around 5 cents per user per month, so performance-obsession is important to have the project sustainable from an operational-costs point of view.

I am happy to discuss options for different approaches.
